src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -486,23 +559,154 @@ InsertValidationResponse insertBufferedRecords(StreamingBuffer streamingBufferTo
     return response;
   }
 
-  /**
-   * Uses {@link Fallback} API to reopen the channel if insertRows throws {@link SFException}.
-   *
-   * <p>We have deliberately not performed retries on insertRows because it might slow down overall
-   * ingestion and introduce lags in committing offsets to Kafka.
-   *
-   * <p>Note that insertRows API does perform channel validation which might throw SFException if
-   * channel is invalidated.
-   *
-   * <p>It can also send errors {@link
-   * net.snowflake.ingest.streaming.InsertValidationResponse.InsertError} in form of response inside
-   * {@link InsertValidationResponse}
-   *
-   * @param buffer buffer to insert into snowflake
-   * @return InsertValidationResponse object sent from insertRows API.
-   */
-  private InsertValidationResponse insertRowsWithFallback(StreamingBuffer buffer) {
+  public class InsertRowsWithRetryResponse {
+    public List<InsertValidationResponse.InsertError> insertErrors;
+    public List<SinkRecord> recordWithError;
+
+    private Map<String, String> extraColumnToType;
+
+    private List<String> nonNullableColumns;
+
+    private boolean isChannelReopened;
+
+    private InsertRowsWithRetryResponse() {
+      this.insertErrors = new ArrayList<>();
+      this.recordWithError = new ArrayList<>();
+      this.isChannelReopened = false;
+    }
+
+    public boolean hasErrors() {
+      return !recordWithError.isEmpty();
+    }
+
+    public void setChannelReopened() {
+      this.isChannelReopened = true;
+    }
+
+    public boolean getChannelReopened() {
+      return this.isChannelReopened;
+    }
+
+    /**
+     * Add rows with their errors to the response. The index of the record matches that of the
+     * error.
+     *
+     * <p>Used when all retries failed
+     */
+    public void addRecordsWithErrorsToResponse(
```

Comment:
can you comment here saying this is also reached when there are no retries (or in non schematization case too?)